rules/handler.go

```diff
@@ -0,0 +1,391 @@
+	DeleteMappingRule(ruleName string) error
+
+	// AppendRollupRule creates a new mapping rule and adds it to this ruleset.
+	AddRollupRule(RollupRuleConfig) error
+
+	// UpdateRollupRule creates a new mapping rule and adds it to this ruleset.
+	UpdateRollupRule(RollupRuleUpdate) error
+
+	// DeleteRollupRule deletes a rollup rule
+	DeleteRollupRule(ruleName string) error
+
+	// Tombstone tombstones this ruleset and all of its rules.
+	Delete() error
+
+	// Revive removes the tombstone from this ruleset. It does not revive any rules.
+	Revive() error
+
+	// Validate checks that this ruleset is good to use. e.g. not tombstoned
+	Validate() error
+
+	// Ruleset returns the latest state of the ruleset in this handler.
+	RuleSet() RuleSet
+}
+
+// ruleSetHandler is a struct that all functionality should be defined on.
+type ruleSetHandler struct {
+	opts HandlerOptions
+	rs   *ruleSet
+}
+
+func (h ruleSetHandler) RuleSet() RuleSet {
+	return h.rs
+}
+
+func (h ruleSetHandler) AddMappingRule(mrc MappingRuleConfig) error {
+	rs := h.rs
+	m, err := rs.getMappingRuleByName(mrc.Name)
+	updateTime := time.Now().UnixNano()
+	cutoverTime := updateTime + int64(h.opts.propagationDelay)
+	if err != nil {
+		if err != errNoSuchRule {
+			return fmt.Errorf(ruleActionErrorFmt, "add", mrc.Name, err)
+		}
+		m, err := newMappingRuleFromFields(
+			mrc.Name,
+			mrc.Filters,
+			mrc.Policies,
+			cutoverTime,
+			rs.tagsFilterOpts,
+		)
+		if err != nil {
+			return fmt.Errorf(ruleActionErrorFmt, "add", mrc.Name, err)
+		}
+		rs.mappingRules = append(rs.mappingRules, m)
+	} else {
+		if err := m.revive(mrc.Name, mrc.Filters, mrc.Policies, cutoverTime, rs.tagsFilterOpts); err != nil {
+			return fmt.Errorf(ruleActionErrorFmt, "revive", mrc.Name, err)
+		}
+	}
+
+	rs.updateTimeStamps(updateTime, cutoverTime)
+	return nil
+}
+
+func (h ruleSetHandler) UpdateMappingRule(mru MappingRuleUpdate) error {
+	rs := h.rs
+	m, err := rs.getMappingRuleByName(mru.Name)
+	if err != nil {
+		return fmt.Errorf(ruleActionErrorFmt, "update", mru.Name, err)
+	}
+
+	updateTime := time.Now().UnixNano()
+	cutoverTime := updateTime + int64(h.opts.propagationDelay)
+	mrc := mru.Config
+	err = m.addSnapshot(mrc.Name, mrc.Filters, mrc.Policies, cutoverTime, rs.tagsFilterOpts)
+	if err != nil {
+		return fmt.Errorf(ruleActionErrorFmt, "update", mru.Name, err)
+	}
+	rs.updateTimeStamps(updateTime, cutoverTime)
+	return nil
+}
+
+func (h ruleSetHandler) DeleteMappingRule(name string) error {
+	rs := h.rs
+	m, err := rs.getMappingRuleByName(name)
+	if err != nil {
+		return fmt.Errorf(ruleActionErrorFmt, "delete", name, err)
+	}
+	updateTime := time.Now().UnixNano()
+	cutoverTime := updateTime + int64(h.opts.propagationDelay)
+	if err := m.markTombstoned(cutoverTime); err != nil {
+		return fmt.Errorf(ruleActionErrorFmt, "delete", name, err)
+	}
+	rs.updateTimeStamps(updateTime, cutoverTime)
+	return nil
+}
+
+func (h ruleSetHandler) AddRollupRule(rrc RollupRuleConfig) error {
```

Comment:
Do we have validation logic that ensures no two rules can have the same name? Also do we have validation logic to ensure no two rollup targets can have the same (name, list rollup tags)?